### ISSUE TITLE
Question about the generate image difference between online system and github code

### ISSUE THREAD

**Issue by dejavulz** [IGLICT/NeRFFaceLighting#8]
Thank you very much for sharing the code and the online demo system. It allowed me to intuitively experience the impressive results of the paper. The multi-view images generated after uploading a photo to the online system are truly stunning!

I tried to reproduce similar results by following the demo.ipynb in the projector directory and inputting new images. However, I encountered a significant difference between the generated images and the original ones — for example, the background becomes blurry and facial features are noticeably altered. In contrast, the images generated by the online system are much more accurate and visually appealing.

May I ask if there are any additional processing steps or settings used in the online system that are not included in the demo.ipynb code? I'd greatly appreciate your guidance on how to achieve similar quality results locally.

following is the code refer to NeRFFaceLighting-main/projector/demo.ipynb：
```python

import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0'
device = "cuda"
# test_import.py
import sys
sys.path.append('/home/dejavu/Code/NeRFFaceLighting-main/projector')

import sys
sys.path.append("./modules/")
sys.path.append("./modules/models/")

import cv2
import dnnlib
import torch
import pickle
import mrcfile
import pyshtools
import numpy as np
from PIL import Image
from tqdm import tqdm
from typing import List, Optional, Tuple, Union

import imageio
import torchvision
from torchvision import transforms
from torchvision.utils import make_grid

from matplotlib import pyplot as plt
from camera_utils import LookAtPoseSampler, FOV_to_intrinsics

import warnings
warnings.filterwarnings("ignore")

@torch.no_grad()
def render_tensor(img: torch.Tensor, normalize: bool = True, nrow: int = 8) -> Image.Image:
    if type(img) == list:
        img = torch.cat(img, dim=0).expand(-1, 3, -1, -1)
    elif len(img.shape) == 3:
        img = img.expand(3, -1, -1)
    elif len(img.shape) == 4:
        img = img.expand(-1, 3, -1, -1)
    
    img = img.squeeze()
    
    if normalize:
        img = img / 2 + .5
    
    if len(img.shape) == 3:
        return Image.fromarray((img.permute(1, 2, 0).cpu().numpy() * 255).astype(np.uint8))
    elif len(img.shape) == 2:
        return Image.fromarray((img.cpu().numpy() * 255).astype(np.uint8))
    elif len(img.shape) == 4:
        return Image.fromarray((make_grid(img, nrow=nrow).permute(1, 2, 0).cpu().numpy() * 255).astype(np.uint8))
def to_tensor(img: Union[Image.Image, np.ndarray], normalize=True) -> torch.Tensor:
    if isinstance(img, Image.Image):
        img = np.array(img)
        if len(img.shape) > 2:
            img = img.transpose(2, 0, 1)
        else:
            img = img[None, ...]
    else:
        if img.shape[0] == img.shape[1]:
            img = img.transpose(2, 0, 1)
    if normalize:
        img = torch.from_numpy(img).to(torch.float32) / 127.5 - 1
    else:
        img = torch.from_numpy(img).to(torch.float32) / 255.
    return img[None, ...].to(device)   
def load_sh(index):
    sh = np.loadtxt(os.path.join("../data/example_light", f"rotate_light_{index}.txt"))
    sh = sh[0:9] * 0.7
    sh = np.reshape(sh, (1, 9)).astype(np.float32)
    sh = torch.from_numpy(sh).to(device)
    return sh

rotating_sh_s = [
    load_sh("00"), 
    load_sh("01"), 
    load_sh("02"), 
    load_sh("03"), 
    load_sh("04"), 
    load_sh("05"), 
    load_sh("06"), 
    load_sh("00"), 
]
from modules.crop import *
from modules.pose import *
from modules.project import *
import sys
sys.argv = [''] 
cropper = Cropper(device)
poser = Poser(device)
projector = Projector(
    device, 
    "../data/NeRFFaceLighting-ffhq-64.pkl", 
    "../data/encoder.pt"
)
from external_dependencies.DPR.DPR import DPR
dpr = DPR(device)
# image = Image.open(os.path.join("./gallery", "example.jpg")).convert("RGB")
image = Image.open("/home/dejavu/Code/NeRFFaceLighting-main/03-01/03-01/Image1392648615840672000.png").convert("RGB")
keypoints = eg3d_detect_keypoints(image)[0]
pred_coeffs = cropper.get_deep3d_coeffs(image, keypoints)
im = cropper.final_crop(image, keypoints)
img = (torch.from_numpy(np.array(im)).float().permute(2, 0, 1)[None, ...]/127.5-1).to(device)
print(img.shape)
pose = poser.get_pose(pred_coeffs)
sh = dpr.extract_lighting(img).squeeze()[None, ...]
cam = torch.cat([torch.from_numpy(pose["pose"]).to(device).reshape(-1, 16), projector.intrinsics.reshape(-1, 9)], dim=1).float()
G_opt, w_opt, sh_opt, w_opt_loss, sh_opt_loss, pti_loss = projector.project(im, cam, sh)
cam2world_pose = LookAtPoseSampler.sample(6*np.pi/12, 1.2*np.pi/2, projector.cam_pivot, radius=projector.cam_radius, device=device)
camera_params = torch.cat([cam2world_pose.reshape(-1, 16), projector.intrinsics.reshape(-1, 9)], 1)
out = G_opt.synthesis(w_opt, projector.mapping_lit(G_opt, sh_opt), camera_params, None, noise_mode="const")
relit_out = G_opt.synthesis(w_opt, projector.mapping_lit(G_opt, rotating_sh_s[0]), camera_params, None, noise_mode="const")
```
This is the original image I use:
![Image](https://github.com/user-attachments/assets/901681c4-e8f7-4327-8adb-4215b193bd52)

The online system generate image is good:

<img width="1575" alt="Image" src="https://github.com/user-attachments/assets/30c1cc40-e9b2-4e9c-8325-8c04cd048a20" />

But I use code to generate image is different:
`render_tensor([to_tensor(im), out["image"].clamp(-1, 1), out["image_albedo"].clamp(-1, 1), relit_out["image"].clamp(-1, 1)])`

![Image](https://github.com/user-attachments/assets/194fac9c-e73f-4225-9577-41163d4f4c05)

Any advice or pointers would be deeply appreciated. Thanks again for open-sourcing this amazing project!

**Comment by RaymondJiangkw** [comment]
Hi! Thanks for your interest. Well, if I remember correctly, there should be no difference between the online system and the demo code. And sorry that I cannot access the code of the online system now for checking... But by looking at the results, I wonder the projection seems a bit wrong. Have u successfully reproduced the projection notebook result?

If u can reproduce that, in that case, I may indeed tune the parameters of projection on the online system (2-3 years ago!), but I don’t remember.

I would suggest you check the PTI projection by looking at the encoded image before finetuning the generator (should be correct if cropping is correct), and then looking at the loss curve when finetuning the generator and adjusting the weight for loss. :)

**Comment by dejavulz** [comment]
Thank you very much for your reply and the suggestions!
I checked the crop image and e4e transform is right：
```python
 if self.encoder_path != None:
            input_img = self.e4e_transform(img).to(self.device)[None, ...]
            input_img_np = input_img.squeeze(0).cpu().numpy()  # 移除batch维度并转换为numpy
            input_img_np = (input_img_np + 1) / 2  # 将范围从[-1,1]转换到[0,1]
            input_img_np = np.transpose(input_img_np, (1, 2, 0))  # 从(C,H,W)转换为(H,W,C)
            np.save('/home/dejavu/Code/NeRFFaceLighting-main/loss_data/input_img.npy', input_img_np)
```
![Image](https://github.com/user-attachments/assets/96c8ab43-bb11-4cab-805d-48edde963e64)
I tried running `projector/demo.ipynb`, and please forgive me for not carefully comparing the results earlier — I realized that even the original code doesn't produce exactly the same results as shown in the images in the repository.Are the model weights I use different?
```python
projector = Projector(
    device, 
    "../data/NeRFFaceLighting-ffhq-64.pkl", 
    "../data/encoder.pt"
)
```
![Image](https://github.com/user-attachments/assets/f9706d14-ed7b-4233-aaf7-8b1b5b5b8ae5)

And I saved the `sh_opt_loss` from the project function in `NeRFFaceLighting-main/projector/modules/project.py`. (Note: the `w_opt_loss` was not iterated since the encoder_path is not none, so it remained unchanged.)

![Image](https://github.com/user-attachments/assets/e4824453-7728-4f12-b671-c6f63047aa83)

By the way，I noticed that other codes require camera parameters to be provided. How can I calculate the camera parameters for the input wild image?

I’m relatively new to this field, so I would really appreciate your insights

**Comment by RaymondJiangkw** [comment]
This is a bit weird. It looks like the PTI doesn’t work. I think it is better to check the image space loss. I will double check when I have time, but I am quite busy recently. So maybe a bit late.

For getting the camera parameters, the code for cropping and estimating the camera parameters is provided. You can find that in the projection demo code.

**Comment by dejavulz** [comment]
Thank you so much for taking the time to reply despite your busy schedule. I’ll check the LPIPS loss as soon as I can, and please feel free to reach out anytime if you discover anything new. Thanks again!

**Comment by dejavulz** [comment]
Thank you very much for your guidance! I found that I had commented out the if not skip_pti block in the `projector.py` file, which caused `PTI` to be skipped entirely. After restoring that part, everything worked as expected.

I have another question I’d like to ask: I need to process every single frame of a video, and according to the `demo.ipynb`, it takes about 2 minutes to process just one frame on an RTX 4090—especially since I need to apply adjustments to each frame individually. For a one-minute video, this becomes prohibitively time-consuming. Do you have any suggestions for improving the efficiency of processing an entire video?

**Comment by RaymondJiangkw** [comment]
Oh, I see. Nice you solve the problem! Yes, you may want to check [our work](https://github.com/GhostCai/PortraitRelighting/tree/main) for real-time performance. :)
